### ISSUE TITLE
False positives after last release 3 days ago

### ISSUE THREAD

**Issue by dereuromark** [vimeo/psalm#1126]
See https://travis-ci.org/cakephp/cakephp/jobs/467742190

ERROR: UndefinedClass - src/Log/LogEngineRegistry.php:82:25 - Type Psr\Log\LoggerInterface cannot be called as a class
            $instance = new $class($settings);

for
```php
    @param string|\Psr\Log\LoggerInterface $class
    ...
        if (is_object($class)) {
            $instance = $class;
        }

        if (!isset($instance)) {
            $instance = new $class($settings);
        }
```

and

ERROR: EmptyArrayAccess - src/View/Helper/HtmlHelper.php:832:21 - Cannot access value on empty array variable $cellOptions
                    $cellOptions['class'] .= ' column-' . $i;

for
```php
                if (isset($cellOptions['class'])) {
                    $cellOptions['class'] .= ' column-' . $i;
```

etc are all invalid and wrong to be reported.

Please be a bit more conservative with the errors.
If not sure, it should rather not voice an error, especially not on a low level.

**Comment by muglug** [comment]
Smaller reproduction: https://getpsalm.org/r/c77e16fd4e

**Comment by muglug** [comment]
That second one is fixed, @dereuromark, but you can make this go away by typing `$line` more strictly

**Comment by muglug** [comment]
I'm going to close this. The error you're seeing with `src/Log/LogEngineRegistry.php` would have always been raised as an issue in Psalm had everything been typed properly (e.g. https://getpsalm.org/r/6051bbe0aa) but it's only being raised in 3.0.3 because Psalm now [checks unions of `mixed` variables](https://github.com/vimeo/psalm/releases/tag/3.0.3) (the same feature caused the bug fixed in https://github.com/vimeo/psalm/commit/15320430dbc6770928374b95a94f067de5ed185b).

I don't have an easy suggestion for you with `src/Log/LogEngineRegistry.php` - I don't know exactly what exactly you allow `$class` to be, but adding `@psalm-param` annotation and using an `if...else` seems the best bet to make Psalm happy: https://getpsalm.org/r/7128aa3ff8

The challenge for Psalm is understanding that `$instance` is only unset

**Comment by muglug** [comment]
While @tomtomau’s right that everything is configurable, you shouldn’t have to change your configuration to adjust for Psalm bugs. 

Currently all releases are checked against Vimeo’s codebase, but there’s clearly a need for more comprehensive pre-release testing. To that end I’ll create some some sort of testbed that includes large Psalm-using open source projects so as to avoid regressions.
